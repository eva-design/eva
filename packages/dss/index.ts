export {
  SchemaType,
  StrictTheme,
  ThemeMappingType,
  ControlMappingType,
  ControlMetaType,
  ScopeMetaType,
  PropertyMetaType,
  AppearanceMetaType,
  VariantMetaType,
  StateMetaType,
  AppearanceMappingType,
  MappingType,
  StateMappingType,
  ParameterType,
  ThemeStyleType,
  ControlThemedStyleType,
  ThemedStyleType,
} from './types/schema';

export { CustomSchemaType } from './types/schema-customization';
