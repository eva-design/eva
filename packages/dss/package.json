{
  "name": "@eva-design/dss",
  "description": "Eva Design System Schema",
  "version": "0.0.1",
  "license": "MIT",
  "author": "akveo <contact@akveo.com>",
  "homepage": "https://github.com/akveo/eva#readme",
  "repository": "git+https://github.com/akveo/eva.git",
  "bugs": {
    "url": "https://github.com/akveo/eva/issues"
  }
}
