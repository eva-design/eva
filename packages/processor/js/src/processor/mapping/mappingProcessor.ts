import {
  ThemeMappingType,
  ControlMappingType,
} from '@eva-design/dss';
import { Processor } from '../processor';
import {
  getComponentVariants,
  getComponentStates,
  SEPARATOR_MAPPING_ENTRY,
} from '../../service';

export interface MappingMetaType {
  name: string;
  appearance: string;
  variants: string[];
  states: string[];
}

export class MappingProcessor implements Processor<ThemeMappingType, MappingMetaType[]> {

  public process(params: ThemeMappingType): MappingMetaType[] {
    return Object.keys(params).reduce((acc: MappingMetaType[], component: string) => {
      return [
        ...acc,
        ...this.getComponentMappingMeta(params, component),
      ];
    }, []);
  }

  private getComponentMappingMeta(mapping: ThemeMappingType, component: string): MappingMetaType[] {
    const componentMapping: ControlMappingType = mapping[component];

    return Object.keys(componentMapping.appearances).map((appearance: string): MappingMetaType => {
      return {
        name: component,
        appearance: appearance,
        variants: this.getComponentVariants(mapping, component),
        states: this.getComponentStates(mapping, component),
      };
    });
  }

  private getComponentVariants(mapping: ThemeMappingType, component: string): string[] {
    const variants: string[][] = getComponentVariants(mapping, component);

    return this.concatComponentVariants([...variants]);
  }

  private getComponentStates(mapping: ThemeMappingType, component: string): string[] {
    const states: string[] = getComponentStates(mapping, component);

    return this.concatComponentStates([...states]);
  }

  private concatComponentVariants(variants: string[][], result: string[] = []): string[] {
    if (variants.length === 0) {
      return result;
    }

    const concat = variants.reduce((acc: string[], current: string[]) => {
      return [...acc, ...this.concatVariantGroups(acc, current)];
    }, variants.shift());

    return this.concatComponentVariants(variants, [...result, ...concat]);
  }

  private concatVariantGroups(lhs: string[], rhs: string[]): string[] {
    return lhs.reduce((acc: string[], lhsValue: string) => {
      const concat = rhs.map(rhsValue => {
        return lhsValue.concat(SEPARATOR_MAPPING_ENTRY, rhsValue);
      });
      return [...acc, ...concat];
    }, []);
  }

  private concatComponentStates(states: string[], result: string[] = []): string[] {
    if (states.length === 0) {
      return result;
    }

    const concat = states.reduce((acc: string[], current: string) => {
      const next = acc.map(value => value.concat(SEPARATOR_MAPPING_ENTRY, current));
      return [...acc, ...next];
    }, [states.shift()]);

    return this.concatComponentStates(states, [...result, ...concat]);
  }
}
