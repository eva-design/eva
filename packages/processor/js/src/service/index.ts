export * from './common';
export * from './mapping';
export * from './style';
