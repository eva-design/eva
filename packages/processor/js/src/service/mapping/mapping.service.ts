import {
  ThemeMappingType,
  ControlMappingType,
  AppearanceMappingType,
  MappingType,
  StateMappingType,
} from '@eva-design/dss';
import { safe } from '../common';

export function getComponentDefaultAppearance(mapping: ThemeMappingType,
                                              component: string): string | undefined {

  const componentMapping = getComponentMapping(mapping, component);

  return safe(componentMapping, (value: ControlMappingType) => {
    const { appearances } = value.meta;

    return Object.keys(appearances).find((appearance: string) => {
      return appearances[appearance].default === true;
    });
  });
}

export function getComponentDefaultVariants(mapping: ThemeMappingType,
                                            component: string): string[] | undefined {

  const componentMapping = getComponentMapping(mapping, component);

  return safe(componentMapping, (value: ControlMappingType) => {
    const { variantGroups } = value.meta;

    return Object.keys(variantGroups).map((group: string) => {
      return Object.keys(variantGroups[group]).find((variant: string) => {
        return variantGroups[group][variant].default === true;
      });
    });
  });
}

export function getComponentDefaultState(mapping: ThemeMappingType,
                                         component: string): string | undefined {

  const componentMapping = getComponentMapping(mapping, component);

  return safe(componentMapping, (value: ControlMappingType) => {
    const { states } = value.meta;

    return Object.keys(states).find((state: string) => {
      return states[state].default === true;
    });
  });
}

export function getComponentVariantGroups(mapping: ThemeMappingType,
                                          component: string): string[] | undefined {

  const componentMapping = getComponentMapping(mapping, component);

  return safe(componentMapping, (value: ControlMappingType) => {
    const { variantGroups } = value.meta;

    return Object.keys(variantGroups);
  });
}

export function getComponentVariants(mapping: ThemeMappingType,
                                     component: string): string[][] | undefined {

  const componentMapping = getComponentMapping(mapping, component);

  return safe(componentMapping, (value: ControlMappingType) => {
    const { variantGroups } = value.meta;

    return Object.keys(variantGroups).map((group: string) => {
      return Object.keys(variantGroups[group]);
    });
  });
}

export function getComponentStates(mapping: ThemeMappingType,
                                   component: string): string[] | undefined {

  const componentMapping = getComponentMapping(mapping, component);

  return safe(componentMapping, (value: ControlMappingType) => {
    const { states } = value.meta;

    return Object.keys(states);
  });
}

export function getStateAppearanceMapping(mapping: ThemeMappingType,
                                          component: string,
                                          appearance: string,
                                          state: string): MappingType | undefined {

  const appearanceMapping = getAppearanceMapping(mapping, component, appearance);

  return safe(appearanceMapping, (value: StateMappingType) => {
    return getStateMapping(value, state);
  });
}

export function getStatelessAppearanceMapping(mapping: ThemeMappingType,
                                              component: string,
                                              appearance: string): MappingType | undefined {

  const appearanceMapping = getAppearanceMapping(mapping, component, appearance);

  return safe(appearanceMapping, (value: StateMappingType) => {
    const { state, ...params } = value;

    return params;
  });
}

export function getStateVariantMapping(mapping: ThemeMappingType,
                                       component: string,
                                       appearance: string,
                                       variant: string,
                                       state: string): MappingType | undefined {

  const variantMapping = getVariantMapping(mapping, component, appearance, variant);

  return safe(variantMapping, (value: MappingType) => {
    return getStateMapping(value, state);
  });
}

export function getStatelessVariantMapping(mapping: ThemeMappingType,
                                           component: string,
                                           appearance: string,
                                           variant: string): MappingType | undefined {

  const variantMapping = getVariantMapping(mapping, component, appearance, variant);

  return safe(variantMapping, (value: MappingType) => {
    const { state, ...params } = value;

    return params;
  });
}

export function getComponentMapping(mapping: ThemeMappingType,
                                    component: string): ControlMappingType | undefined {

  return mapping[component];
}

function getAppearance(mapping: ThemeMappingType,
                       component: string,
                       appearance: string): AppearanceMappingType | undefined {

  const componentMapping = getComponentMapping(mapping, component);

  return safe(componentMapping, (value: ControlMappingType) => {
    return value.appearances[appearance];
  });
}

function getAppearanceMapping(mapping: ThemeMappingType,
                              component: string,
                              appearance: string): StateMappingType | undefined {

  const appearanceConfig = getAppearance(mapping, component, appearance);

  return safe(appearanceConfig, (value: AppearanceMappingType) => {
    return value.mapping;
  });
}

function getVariantMapping(mapping: ThemeMappingType,
                           component: string,
                           appearance: string,
                           variant: string): MappingType | undefined {

  const appearanceConfig = getAppearance(mapping, component, appearance);

  return safe(appearanceConfig, (value: AppearanceMappingType) => {
    return safe(value.variantGroups, (groupValue) => {

      const groupName = Object.keys(groupValue).find((group: string) => {
        return groupValue[group][variant] !== undefined;
      });

      return safe(groupName, (groupNameValue) => {
        return groupValue[groupNameValue][variant];
      });
    });
  });
}

function getStateMapping(mapping: StateMappingType,
                         state: string): MappingType | undefined {

  return safe(mapping.state, (value) => {
    return value[state];
  });
}
