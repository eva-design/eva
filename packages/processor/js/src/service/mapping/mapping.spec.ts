import { MappingType } from '@eva-design/dss';
import { mapping } from './mapping.spec.config';
import * as Service from './mapping.service';

describe('@mapping: service methods checks', () => {

  describe('* appearance', () => {

    describe('* default', () => {

      const appearance: string = 'filled';

      it('* stateless', () => {
        const value: MappingType = Service.getStatelessAppearanceMapping(
          mapping,
          'Button',
          appearance,
        );

        expect(value).toMatchSnapshot();
      });

      describe('* state', () => {

        it('* active', () => {
          const value: MappingType = Service.getStateAppearanceMapping(
            mapping,
            'Button',
            appearance,
            'active',
          );

          expect(value).toMatchSnapshot();
        });

        it('* undefined', () => {
          const value: MappingType = Service.getStateAppearanceMapping(
            mapping,
            'Button',
            appearance,
            'undefined',
          );

          expect(value).toBeUndefined();
        });

      });

      describe('* variant', () => {

        it('* success', () => {
          const value: MappingType = Service.getStatelessVariantMapping(
            mapping,
            'Button',
            appearance,
            'success',
          );

          expect(value).toMatchSnapshot();
        });

        it('* undefined', () => {
          const value = Service.getStatelessVariantMapping(
            mapping,
            'Button',
            appearance,
            'undefined',
          );

          expect(value).toBeUndefined();
        });

        describe('* state', () => {

          it('* active', () => {
            const value: MappingType = Service.getStateVariantMapping(
              mapping,
              'Button',
              appearance,
              'success',
              'active',
            );

            expect(value).toMatchSnapshot();
          });

          it('* undefined', () => {
            const value: MappingType = Service.getStateVariantMapping(
              mapping,
              'Button',
              appearance,
              'success',
              'undefined',
            );

            expect(value).toBeUndefined();
          });

        });

      });

    });

    describe('* custom', () => {

      const appearance: string = 'outline';

      it('* stateless', () => {
        const value = Service.getStatelessAppearanceMapping(
          mapping,
          'Button',
          appearance,
        );

        expect(value).toMatchSnapshot();
      });

      describe('* state', () => {

        it('* active', () => {
          const value = Service.getStateAppearanceMapping(mapping,
            'Button',
            appearance,
            'active',
          );

          expect(value).toMatchSnapshot();
        });

        it('* undefined', () => {
          const value = Service.getStateAppearanceMapping(
            mapping,
            'Button',
            appearance,
            'undefined',
          );

          expect(value).toBeUndefined();
        });

      });

      describe('* variant', () => {

        it('* success', () => {
          const value = Service.getStatelessVariantMapping(
            mapping,
            'Button',
            appearance,
            'success',
          );

          expect(value).toMatchSnapshot();
        });

        it('* undefined', () => {
          const value = Service.getStatelessVariantMapping(
            mapping,
            'Button',
            appearance,
            'undefined',
          );

          expect(value).toMatchSnapshot();
        });

        describe('* state', () => {

          it('* active', () => {
            const value = Service.getStateVariantMapping(
              mapping,
              'Button',
              appearance,
              'success',
              'active',
            );

            expect(value).toMatchSnapshot();
          });

          it('* undefined', () => {
            const value = Service.getStateVariantMapping(
              mapping,
              'Button',
              appearance,
              'success',
              'undefined',
            );


            expect(value).toBeUndefined();
          });

        });

      });

    });

    describe('* undefined', () => {

      const appearance = 'undefined';

      it('* stateless', () => {
        const value = Service.getStatelessAppearanceMapping(
          mapping,
          'Button',
          appearance,
        );

        expect(value).toBeUndefined();
      });

      describe('* state', () => {

        it('* active', () => {
          const value = Service.getStateAppearanceMapping(
            mapping,
            'Button',
            appearance,
            'active',
          );

          expect(value).toBeUndefined();
        });

        it('* undefined', () => {
          const value = Service.getStateAppearanceMapping(
            mapping,
            'Button',
            appearance,
            'undefined',
          );

          expect(value).toBeUndefined();
        });

      });

      describe('* variant', () => {

        it('* success', () => {
          const value = Service.getStatelessVariantMapping(
            mapping,
            'Button',
            appearance,
            'success',
          );

          expect(value).toBeUndefined();
        });

        it('* undefined', () => {
          const value = Service.getStatelessVariantMapping(
            mapping,
            'Button',
            appearance,
            'undefined',
          );

          expect(value).toBeUndefined();
        });

        describe('* state', () => {

          it('* active', () => {
            const value = Service.getStateVariantMapping(
              mapping,
              'Button',
              appearance,
              'success',
              'active',
            );

            expect(value).toBeUndefined();
          });

          it('* undefined', () => {
            const value = Service.getStateVariantMapping(
              mapping,
              'Button',
              appearance,
              'success',
              'undefined',
            );

            expect(value).toBeUndefined();
          });

        });

      });

    });

  });

  describe('* variants groups getting', () => {

    it('* defined component', () => {
      const value: string[] = Service.getComponentVariantGroups(
        mapping,
        'Button',
      );

      expect(value).toEqual(['status', 'size']);
    });

    it('* undefined component', () => {
      const value: string[] | undefined = Service.getComponentVariantGroups(mapping, 'Undefined');

      expect(value).toBeUndefined();
    });

  });

});
