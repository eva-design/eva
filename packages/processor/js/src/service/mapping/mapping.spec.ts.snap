// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@mapping: service methods checks * appearance * custom * state * active 1`] = `undefined`;

exports[`@mapping: service methods checks * appearance * custom * stateless 1`] = `
Object {
  "borderWidth": 2,
}
`;

exports[`@mapping: service methods checks * appearance * custom * variant * state * active 1`] = `
Object {
  "backgroundColor": "transparent",
  "borderColor": "#43A047",
  "iconTintColor": "#43A047",
  "textColor": "#43A047",
}
`;

exports[`@mapping: service methods checks * appearance * custom * variant * success 1`] = `
Object {
  "backgroundColor": "transparent",
  "borderColor": "#4CAF50",
  "iconTintColor": "#4CAF50",
  "textColor": "#4CAF50",
}
`;

exports[`@mapping: service methods checks * appearance * custom * variant * undefined 1`] = `undefined`;

exports[`@mapping: service methods checks * appearance * default * state * active 1`] = `undefined`;

exports[`@mapping: service methods checks * appearance * default * stateless 1`] = `
Object {
  "borderRadius": 6,
  "iconMarginHorizontal": 4,
  "iconTintColor": "#ffffff",
  "textColor": "text-primary-inverse",
  "textFontWeight": 800,
}
`;

exports[`@mapping: service methods checks * appearance * default * variant * state * active 1`] = `
Object {
  "backgroundColor": "#43A047",
}
`;

exports[`@mapping: service methods checks * appearance * default * variant * success 1`] = `
Object {
  "backgroundColor": "#4CAF50",
}
`;
